src/Elastic.Apm/Api/IExecutionSegment.cs

```diff
@@ -40,7 +40,7 @@ public interface IExecutionSegment
 		/// For example <code>foo.bar</code> will be stored as <code>foo_bar</code> in Elasticsearch.
 		/// </summary>
 		/// <exception cref="ArgumentException"><c>null</c> as key is not allowed.</exception>
-		Dictionary<string, string> Labels { get; }
+		Dictionary<string, object> Labels { get; }
```

Comment:
This isn't very nice, but I want to keep backwards compatibility - this way things like (current API usage) this will still compile and work:

```
ISpan.Labels["foo"] = "bar"
``` 

Long term I feel it'd be better to just depreciate the `Dictionary` from the API.

Reply:
Could a `Tag` (from what I see, tag is what labels really are?) type be introduced, that constrains the inputs only to 

- string
- bool
- double
- int
- long
- decimal

For example,

```c#
[JsonConverter(typeof(TagConverter))]
public class Tag
{
	internal object _object;
	
	public Tag(string value) => _object = value;
	
	public Tag(bool value) => _object = value;
	
	public Tag(double value) => _object = value;
	
	public Tag(int value) => _object = value;
	
	public Tag(long value) => _object = value;
	
	public Tag(decimal value) => _object = value;
	
	public static implicit operator Tag(string value) => new Tag(value);
	
	public static implicit operator Tag(bool value) => new Tag(value);
	
	public static implicit operator Tag(double value) => new Tag(value);
	
	public static implicit operator Tag(int value) => new Tag(value);
	
	public static implicit operator Tag(long value) => new Tag(value);
	
	public static implicit operator Tag(decimal value) => new Tag(value);
}

public class TagConverter : JsonConverter
{
	public override bool CanConvert(Type objectType) => typeof(Tag).IsAssignableFrom(objectType);

	public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
	{
		switch(reader.TokenType)
		{ 
			case JsonToken.Boolean:
				return new Tag((bool)reader.Value);
			case JsonToken.Float:
				return new Tag((double)reader.Value);
			case JsonToken.Integer:
				return new Tag((long)reader.Value);
			case JsonToken.String:
				return new Tag((string)reader.Value);
			default:
				throw new JsonException($"cannot deserialize {nameof(Tag)} from token type {reader.TokenType}");
		}
	}

	public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer) =>
		writer.WriteValue(((Tag)value)._object);
}
```

This would maintain API compatibility and also constrain tags/labels to the types supported by APM server

Reply:
I really liked this idea and I also implemented it (as discussed on zoom I named the class `Label`, but I use the same idea -  and I pushed it to this PR.

I ran into one (I think) fairly big problem: This indeed makes it a nice API and it's backward-compatible with things like `ISpan.Labels["foo"] = "bar"`, but once you want to read the value it becomes a breaking change:

Prior to this, here is how you read the value:
```
var myValue = `ISpan.Labels["foo"] ; // here `myValue` is a string
```

But once we add the `Label` type, `myValue` won't be a `string` anymore, it'll be a `Label` and in order to read the value you need to write:

```
var myValue = `ISpan.Labels["foo"].Value
```

This is definitely not the standard use-case (I expect users just to write into the `Labels` and never really read it in the code later), nevertheless this was totally possible with the old API and we'd break that.  ...and this change also breaks our tests from the same reason, so we also need to assert with `.Value` once we introduce the new `Label` type.

So, right now the suggestion is contained in the PR, but I have slight preference to not add this class because of this problem. Sad to not have the cool API that'd promote the possible types, but I'd rather not break the reading use case.  Any idea or thoughts on this?

Reply:
One could look at it from the perspective that the agent's API does not conform to the APM server spec for labels, because it only supports strings, and thus this is a bug that should be addressed. It is unfortunate that the fix is a breaking change.

I think a working around of this breaking change is probably not worth the effort and adds more complexity to the API. For example, an implicit operator of `Label` -> `string` could be added

```c#
public class Label
{
    public static implicit operator string(Label value) => value.Value.ToString();
}
```

`myValue` would still be of type `Label`, but would convert to `string` in cases where a `string` value is expected e.g.

```c#
void Main()
{
	var foo = new Foo();
	
	foo.Labels.Add("label_1", "one");
	foo.Labels.Add("label_2", 2);
	
        // conversion to string on assignment
	string string1 = foo.Labels["label_1"];
	Write(string1);
	
        // conversion to string on pass to method
	var label1 = foo.Labels["label_1"];
	Write(label1);
}

public void Write(string value) => Console.WriteLine(value);

public class Foo
{
	private Dictionary<string, Label> _labels = new Dictionary<string, Label>();	
	public Dictionary<string, Label> Labels => _labels;
}
```

What makes me uncomfortable with this approach is that any value `Label` holds would be converted to string, rather than handled as the type it is.

I personally think it would be better to introduce the breaking change, and bump the major version, if the value of supporting multiple types for label values is important; I suspect that this might be useful to have for performing certain types of queries such as range queries on numeric values.

Reply:
We had a chat about this over zoom. We concluded that actually all the suggestions are breaking changes (no matter if it's  `Dictionary<string,object>` or `Dictionary<string,Label>`, so we decided that I'll do another round on this and try to keep the API as `Dictionary<string,string>` - that's to only way to be sure we don't break user code.

I implemented this and pushed to this PR.